src/buf/fixed/pool.rs

```diff
@@ -223,164 +232,57 @@ impl FixedBufPool {
-        // that has the same length.
-        let iovec = unsafe { self.raw_bufs.as_ptr().add(index).read() };
-        debug_assert_eq!(iovec.iov_len, cap);
-        Some(CheckedOutBuf {
-            iovec,
-            init_len,
-            index: index as u16,
-        })
-    }
-
-    fn check_in_internal(&mut self, index: u16, init_len: usize) {
-        let cap = self.iovecs()[index as usize].iov_len;
-        let state = &mut self.states[index as usize];
-        debug_assert!(
-            matches!(state, BufState::CheckedOut),
-            "the buffer must be checked out"
-        );
-
-        // Link the buffer as the new head of the free list for its capacity.
-        // Recently checked in buffers will be first to be reused,
-        // improving cache locality.
-        let next = self.free_buf_head_by_cap.insert(cap, index);
-
-        *state = BufState::Free { init_len, next };
-    }
-}
-
-impl FixedBuffers for Inner {
-    fn iovecs(&self) -> &[iovec] {
-        // Safety: the raw_bufs pointer is valid for the lifetime of self,
-        // the length of the states array is also the length of buffers array
-        // by construction.
-        unsafe { slice::from_raw_parts(self.raw_bufs.as_ptr(), self.states.len()) }
+        // Poll for a buffer, engaging the `Notify` machinery.
+        self.next_when_notified(cap, notify).await
     }
 
-    unsafe fn check_in(&mut self, index: u16, init_len: usize) {
-        self.check_in_internal(index, init_len)
-    }
-}
+    #[cold]
+    async fn next_when_notified(&self, cap: usize, notify: Arc<Notify>) -> FixedBuf {
+        let notified = notify.notified();
+        pin!(notified);
+        loop {
+            // In the single-threaded case, no buffers could get checked in
+            // between us calling `try_next` and here, so we can't miss a wakeup.
+            notified.as_mut().await;
 
-impl Drop for Inner {
-    fn drop(&mut self) {
-        let iovecs = unsafe {
-            Vec::from_raw_parts(self.raw_bufs.as_ptr(), self.states.len(), self.orig_cap)
-        };
-        for (i, iovec) in iovecs.iter().enumerate() {
-            match self.states[i] {
-                BufState::Free { init_len, next: _ } => {
-                    let ptr = iovec.iov_base as *mut u8;
-                    let cap = iovec.iov_len;
-                    let v = unsafe { Vec::from_raw_parts(ptr, init_len, cap) };
-                    mem::drop(v);
-                }
-                BufState::CheckedOut => unreachable!("all buffers must be checked in"),
+            if let Some(data) = self.inner.borrow_mut().try_next(cap) {
+                // Safety: the validity of buffer data is ensured by
+                // plumbing::Pool::try_next
+                let buf = unsafe { FixedBuf::new(Rc::clone(&self.inner) as _, data) };
+                return buf;
             }
+
+            // Reset the `Notified` future in case another call to `try_next`
+            // got the buffer before us.
```

Comment:
Maybe the comment could add that the permit may have been held even though there was no buffer, which is likely once the exhaustion case has gone away and then comes back.

Reply:
https://github.com/tokio-rs/tokio-uring/pull/199/commits/d83eacb9a484be4cb9567a9977ea117fa833fc56